README.md
# Infosimples::Data

Ruby API for Infosimples::Data (https://infosimples.com/).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'infosimples-data'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install infosimples-data

## Usage

1. **Create a client**

  ```ruby
  require 'infosimples/data'
  client = Infosimples::Data::Client.new('my_token')
  ```

2. **Automate a service**

  ```ruby
  response = client.automate('correios/cep', cep: '01311915')
  response['code']                  # 200
  response['code_message']          # "A consulta foi realizada com sucesso e retornou um resultado."
  response['data']                  # Information about the postal code
  response['receipt']['sites_urls'] # Links for the original HTMLs.

  client.download_sites_urls(response) # Array with HTML bodies for `sites_urls`.
  ```

  > Check the full response in our documentation:
  > https://data.infosimples.com/docs

3. **Billing**

  ```ruby
  billing = client.billing          # Array of Hashes with:
  billing.first['name']             # - Name of the Token
  billing.first['quantity']         # - Number of automation requests the token has made
  billing.first['credits']          # - How many credits the token has used
  ```

4. **Pricing**

  ```ruby
  pricing = client.pricing          # Array of Hashes with:
  pricing.first['service']          # - Name of the service that can be automated
  pricing.first['credits']          # - How many credits this service uses
  ```

5. **Response codes**

  ```ruby
  # Check a response code:
  response['code'] == Infosimples::Data::CODE[:single_result]

  # List of codes
  {
    # success
    single_result:          200,
    multiple_results:       201,

    # error
    unexpected_error:       600,
    unauthorized:           601,
    invalid_service:        602,
    forbidden:              603,
    invalid_request:        604,
    timeout:                605,
    empty_parameters:       606,
    invalid_parameters:     607,
    refused_parameters:     608,
    attempts_exceeded:      609,
    failed_captcha:         610,
    incomplete_data:        611,
    inexistent:             612,
    blocked_request:        613,
    try_again:              614,
    source_unavailable:     615,
    source_error:           616,
    service_overloaded:     617,
    rate_limit_exceeded:    618,
    converted_parameters:   619,
    permanent_error:        620,
    receipt_error:          621,
  }
  ```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/infosimples/infosimples-data.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# CHANGELOG

### [1.9.3] - 2018-10-08

# Added
- GEM created


spec/spec_helper.rb
require "bundler/setup"
require "infosimples/data"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end

require 'yaml'
SECRETS = YAML.load_file('spec/secrets.yml')


spec/infosimples/data_spec.rb
RSpec.describe Infosimples::Data do
  it "has a version number" do
    expect(Infosimples::Data::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


spec/infosimples/client_spec.rb
client   = Infosimples::Data::Client.new(SECRETS['token'])
service  = SECRETS['service']
args     = SECRETS['args']
automate_response = client.automate(service, args)
billing_response  = client.billing
pricing_response  = client.pricing

RSpec.describe Infosimples::Data::Client do
  describe '#automate' do
    it { expect(automate_response['code']).to eq Infosimples::Data::CODE[:single_result] }
    it { expect(automate_response).to have_key('code_message') }
    it { expect(automate_response['header']['service']).to eq service }
    it { expect(automate_response['header']).to have_key('parameters') }
    it { expect(automate_response['header']).to have_key('client') }
    it { expect(automate_response['header']).to have_key('token') }
    it { expect(automate_response['header']).to have_key('billable') }
    it { expect(automate_response['header']).to have_key('credits') }
    it { expect(automate_response['header']).to have_key('cache_hit') }
    it { expect(automate_response['header']).to have_key('elapsed_time_in_milliseconds') }
    it { expect(automate_response).to have_key('data') }
    it { expect(automate_response['receipt']).to have_key('sites_urls') }
  end

  describe '#billing' do
    it { expect(billing_response).to be_instance_of(Array) }
    it { expect(billing_response.sample).to have_key('name') }
    it { expect(billing_response.sample).to have_key('quantity') }
    it { expect(billing_response.sample).to have_key('credits') }
  end

  describe '#pricing' do
    it { expect(pricing_response).to be_instance_of(Array) }
    it { expect(pricing_response.sample).to have_key('service') }
    it { expect(pricing_response.sample).to have_key('credits') }
  end
end


lib/infosimples/data/http.rb
module Infosimples::Data
  # Infosimples::Data::HTTP exposes common HTTP routines that can be used by the
  # Infosimples::Data API client.
  #
  class HTTP
    # Perform an HTTP request with support for multipart requests.
    #
    # @param [Hash] options Options hash.
    # @param options [String] url      URL to be requested.
    # @param options [Symbol] method   HTTP method (:get, :post, :multipart).
    # @param options [Hash]   payload  Data to be sent through the HTTP request.
    # @param options [Integer] http_timeout HTTP open/read timeout in seconds.
    #
    # @return [String] Response body of the HTTP request.
    #
    def self.request(options = {})
      uri     = URI(options[:url])
      method  = options[:method] || :get
      payload = options[:payload] || {}
      timeout = options[:http_timeout]
      headers = { 'User-Agent' => Infosimples::Data::USER_AGENT }

      case method
      when :get
        uri.query = URI.encode_www_form(payload)
        req = Net::HTTP::Get.new(uri.request_uri, headers)

      when :post
        req = Net::HTTP::Post.new(uri.request_uri, headers)
        req.set_form_data(payload)

      when :multipart
        req = Net::HTTP::Post.new(uri.request_uri, headers)
        boundary, body = prepare_multipart_data(payload)
        req.content_type = "multipart/form-data; boundary=#{boundary}"
        req.body = body

      else
        fail Infosimples::Data::ArgumentError, "Illegal HTTP method (#{method})"
      end

      http = Net::HTTP.new(uri.hostname, uri.port)
      http.use_ssl = true if (uri.scheme == 'https')
      http.open_timeout = timeout
      http.read_timeout = timeout + 10
      http.max_retries = 0 # default is max_retries = 1
      res = http.request(req)
      res.body

    rescue Net::OpenTimeout, Net::ReadTimeout
      raise Infosimples::Data::Timeout
    end

    # Prepare the multipart data to be sent via a :multipart request.
    #
    # @param [Hash] payload Data to be prepared via a multipart post.
    #
    # @return [String, String] Boundary and body for the multipart post.
    #
    def self.prepare_multipart_data(payload)
      boundary = 'randomstr' + rand(1_000_000).to_s # a random unique string

      content = []
      payload.each do |param, value|
        content << '--' + boundary
        content << "Content-Disposition: form-data; name=\"#{param}\""
        content << ''
        content << value
      end
      content << '--' + boundary + '--'
      content << ''

      [boundary, content.join("\r\n")]
    end
  end
end


lib/infosimples/data/client.rb
module Infosimples::Data
  # Infosimples::Data::Client is a client for the Infosimples Data API.
  class Client
    BASE_URL = 'https://data.infosimples.com/api/v1/:service.json'
    ENCRYPTABLE_ARGS = ['pkcs12_cert', 'pkcs12_pass', 'pass'].map {
      |arg| [arg, true]
    }.to_h

    attr_accessor :token, :timeout, :max_age

    # Create a Infosimples::Data API client.
    #
    # @param [String] token    Your access token.
    # @param [Hash]   options  Options hash.
    # @option options [Integer] :timeout (120)   Seconds before giving up of an
    #                                            automation being completed.
    # @option options [Integer] :max_age (86400) Duration in seconds for a
    #                                            cached automation to be
    #                                            allowed.
    #
    # @return [Infosimples::Data::Client] A Client instance.
    def initialize(token, options = {})
      self.token    = token
      self.timeout  = options[:timeout] || 120   # 120 seconds
      self.max_age  = options[:max_age] || 86400 # 24 hours in seconds
    end

    # Automate a service.
    #
    # @param [String] service   Service you want to automate.
    # @param [Hash]   args      Arguments to be used in the service automation.
    #
    # @return [Hash] Response according to https://data.infosimples.com/docs.
    def automate(service, args = {})
      args.keys.each do |key|
        if ENCRYPTABLE_ARGS[key.to_s]
          args[key] = encrypt(args[key])
        end
      end
      request(service, :multipart, args)
    end

    # Get billing statistics from your account.
    #
    # @return [Array] Billing statistics per token.
    def billing
      request('billing', :get)
    end

    # Get prices for each service.
    #
    # @return [Array] Service with price.
    def pricing
      request('pricing', :get)
    end

    # Download sites_urls from response.
    #
    # @param [Hash] response  Response returned by #automate.
    #
    # @return [Array] HTML bodies from sites_urls.
    def download_sites_urls(response)
      return [] if !response.is_a?(Hash) ||
                   (sites_urls = response.dig('receipt', 'sites_urls')).nil?
      sites_urls.map do |url|
        Infosimples::Data::HTTP.request(url: url, http_timeout: 30)
      end
    end

    private

    # Apply symmetric encryption to data.
    #
    # @param [String] data The original data that will be encrypted.
    #
    # @return [String] encrypted data.
    def encrypt(data)
      @crypt ||= Infosimples::Data::SymmetricEncryption.new(token)
      @crypt.encrypt(data)
    end

    # Perform an HTTP request to the Infosimples Data API.
    #
    # @param [String] service  API method name.
    # @param [Symbol] method   HTTP method (:get, :post, :multipart).
    # @param [Hash]   payload  Data to be sent through the HTTP request.
    #
    # @return [Hash] Parsed JSON from the API response.
    #
    def request(service, method = :get, payload = {})
      res = Infosimples::Data::HTTP.request(
        url: BASE_URL.gsub(':service', service),
        http_timeout: timeout,
        method: method,
        payload: payload.merge(
          token:   token,
          timeout: timeout,
          max_age: max_age,
          header:  1
        )
      )
      JSON.parse(res)
    end
  end
end


lib/infosimples/data/code.rb
module Infosimples::Data
  CODE = {
    # success
    single_result:          200,
    multiple_results:       201,

    # error
    unexpected_error:       600,
    unauthorized:           601,
    invalid_service:        602,
    forbidden:              603,
    invalid_request:        604,
    timeout:                605,
    empty_parameters:       606,
    invalid_parameters:     607,
    refused_parameters:     608,
    attempts_exceeded:      609,
    failed_captcha:         610,
    incomplete_data:        611,
    inexistent:             612,
    blocked_request:        613,
    try_again:              614,
    source_unavailable:     615,
    source_error:           616,
    service_overloaded:     617,
    rate_limit_exceeded:    618,
    converted_parameters:   619,
    permanent_error:        620,
    receipt_error:          621,
  }
end
